### PR TITLE
Add CloudBuild CI configs

### PR DESCRIPTION
**Pull request by AlCutter**
Adds CloudBuild configs for doing CI tests for PRs.

#2298

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

**Comment by AlCutter** [comment]
Docker tests still need looking at, but I think I'll take that on in a follow-up, and there are issues with etcd which we can look at afterwards too.

File: storage/testdb/testdb.go [hunk 1]
```diff
@@ -34,14 +35,40 @@ import (
...
 
+// mysqlURI returns the connection URI to use for tests.
+// It will prefer the value explicitly set by the --test_mysql_uri flag, if
+// that is unset it will fall back to the value in the MYSQL_URI env var, and
+// if that is unset, then it'll try the DefaultMySQLURI value.
+//
+// Only a subset of the suite of tests in this repo require a database it's
+// not possible to blanket apply the test_mysql_uri flag, and maintaining a separate
+// list of tests which needs DB connectivity is brittle, so this fallback approach
+// means we can provide the connection info in a more ambient fashion.
```
Comment (mhutchinson) [inline storage/testdb/testdb.go:58]: This really makes me sad and a little frightened. Can we just remove the flag and say that it always uses the env variable to at least remove one layer of magic from this?

It really feels like the root problem is that we're creating multiple different mysql databases, yet wanting to use `testdb` like there is a global static singleton. Trying to solve this is going to be a whole new adventure though.

**Review by mhutchinson — COMMENTED** [review]
There's a lot going on here. I don't think we can submit this while the docker tests and etcd tests are broken. Can we update the PR description to have checkboxes for doing this?

Otherwise I think it's looking reasonable.

File: cloudbuild.yaml [hunk 2]
```diff
@@ -6,8 +15,171 @@ substitutions:
...
+  name: 'gcr.io/${PROJECT_ID}/trillian_testbase'
+  entrypoint: ./integration/cloudbuild/run_presubmit.sh
+  args:
+    - --no-linters
+    - --no-generate
+  env:
+    - GOFLAGS=-race --tags=pkcs11
+    - GO_TEST_TIMEOUT=20m
+  waitFor:
+    - presubmit
```
Comment (pavelkalinnikov) [inline cloudbuild.yaml:117]: There is a lot going on here and below. Should we attempt to add one build at a time, for easier reviewability?

File: cloudbuild.yaml [hunk 3]
```diff
@@ -1,3 +1,12 @@
+#############################################################################
```
Comment (pavelkalinnikov) [inline cloudbuild.yaml:1]: The other 2 files ("cloudbuild_master.yaml" and "cloudbuild_tag.yaml") have almost zero comments, and don't explain why they are needed. Is there a doc somewhere?

File: integration/cloudbuild/run_integration.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# Wait for mysql instance to be ready
+while ! mysql --protocol=TCP --host=${MYSQL_HOST} --port=${MYSQL_PORT} --user=root -pbananas -e quit ; do
+ sleep 5
+done
+
+export MYSQL_URI="${MYSQL_USER}:${MYSQL_PASSWORD}@tcp(${MYSQL_HOST}:${MYSQL_PORT})/${MYSQL_DATABASE}"
+
+# If the test will use etcd, then install etcd + tools.
+if [ "${ETCD_DIR}" != "" ]; then
+    go install go.etcd.io/etcd go.etcd.io/etcd/etcdctl github.com/fullstorydev/grpcurl/cmd/grpcurl
```
Comment (pavelkalinnikov) [inline integration/cloudbuild/run_integration.sh:25]: This installs the "latest" version of etcd, doesn't it? Should it fetch the version compatible with Trillian?

File: integration/functions.sh [hunk 5]
```diff
@@ -128,6 +129,11 @@ log_prep_test() {
   local logsigner_opts=''
   local has_etcd=0
 
+  if [[ "${MYSQL_URI}" != "" ]]; then
+    logserver_opts+=" --mysql_uri=${MYSQL_URI}"
+    logsigner_opts+=" --mysql_uri=${MYSQL_URI}"
+  fi
+
   # Start a local etcd instance (if configured).
   if [[ -x "${ETCD_DIR}/etcd" ]]; then
```
Comment (pavelkalinnikov) [inline integration/functions.sh:138]: Obsolete after rebasing on https://github.com/google/trillian/pull/2295?

File: cloudbuild.yaml (hunk 3 above)
Comment (AlCutter) [reply cloudbuild.yaml:1]: They manage/test kube configs for the CI/CD env, I don't know if there are docs but we should make some if not.

File: integration/cloudbuild/prepare.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+set -x
+set -e
+
+git clone --depth=1 https://github.com/googleapis/googleapis.git "$GOPATH/src/github.com/googleapis/googleapis"
+go install \
+    github.com/golang/protobuf/proto \
```
Comment (pavelkalinnikov) [inline integration/cloudbuild/prepare.sh:7]: Again, this installs some "latest" version (master?), which might be why there is an etcd/protobuf panic.

File: integration/cloudbuild/testbase/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+RUN curl -sfL https://golang.org/dl/go1.15.7.linux-amd64.tar.gz | tar -xzf - -C /usr/local
+ENV PATH /usr/local/go/bin:$PATH
+
+RUN curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.18.0
+RUN mkdir protoc && \
+    (cd protoc && \
+    PROTOC_VERSION="3.11.4" && \
+    PROTOC_ZIP="protoc-${PROTOC_VERSION}-linux-x86_64.zip" && \
+    wget "https://github.com/google/protobuf/releases/download/v${PROTOC_VERSION}/${PROTOC_ZIP}" && \
```
Comment (pavelkalinnikov) [inline integration/cloudbuild/testbase/Dockerfile:21]: This corresponds to this repo: https://github.com/protocolbuffers/protobuf

The "go import" in the other file imports from a different repo: https://github.com/golang/protobuf.

I guess, we need to make sure they are compatible. I.e. we need to "go import" from a version not far off in time from protoc v3.11.4.

File: integration/cloudbuild/run_integration.sh (hunk 4 above)
Comment (pavelkalinnikov) [reply integration/cloudbuild/run_integration.sh:25]: I think we should probably use one of releases: https://github.com/etcd-io/etcd/releases

File: integration/cloudbuild/testbase/Dockerfile (hunk 7 above)
Comment (AlCutter) [reply integration/cloudbuild/testbase/Dockerfile:21]: Hmm, I think the first repo gives us the `protoc` command, but the second is the golang specific bits (including `protoc-gen-go`).  You may be onto something with the golang/protobuf version but the `pb/protobuf` version is as it was in the [travis config](https://github.com/google/trillian/blob/master/.travis.yml#L40)

File: integration/cloudbuild/run_presubmit.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+trap "docker-compose -p ${HOSTNAME} -f ./integration/cloudbuild/docker-compose-mysql.yaml down" EXIT
+
+# Wait for mysql instance to be ready
+while ! mysql --protocol=TCP --host=${MYSQL_HOST} --port=${MYSQL_PORT} --user=root -pbananas -e quit ; do
+ sleep 5
+done
+
+# Presumbits need a user with CREATE DATABASE grants since they create temporary databases.
+# For the same reason, this is a URI prefix - tests will add DB names to the end.
+export TEST_MYSQL_URI="root:bananas@tcp(${MYSQL_HOST}:${MYSQL_PORT})/"
```
Comment (pavelkalinnikov) [inline integration/cloudbuild/run_presubmit.sh:18]: Remove the trailing slash consistently? So that "mysqlURI()+"/"+name" remains valid.

**Comment by AlCutter** [comment]
Subsumed by #2296

**Pull request closed without merging**